Introduce debugging APIs for structural inspection of an application

mgechev opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [x] Feature

## What is the new behavior?

Introduces new debugging utilities for structural inspection of an application. The PR expands the `ng` namespace and adds a new `getDirectiveMetadata` method that returns the metadata for a specified directive or component instance.

Additionally, the PR optimizes the `getDirectives` function by:

- Returning an empty array if it's called with a text node
- It invokes `loadLContext` with second argument `false` so it doesn't throw an error

## Does this PR introduce a breaking change?

- [x] Yes, no longer throw an error (when `ng.getDirectives` is called) in case there are no directives on the element 
- [ ] No

mgechev committed:
feat(core): introduce getDirectiveMetadata global debugging utility

This commit introduces a global debugging method
`ng.getDirectiveMetadata` which returns the metadata for a directive or
component instance.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -6,10 +6,13 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {DebugNode as PublicDebugNode, getDebugNode__POST_R3__} from '../../debug/debug_node';
```
I'd propose keeping `DebugNode` here and rename the existing `DebugNode` symbol to `IDebugNode` or something like that. Motivation: this type would be used in `asDebugNode` function, which might potentially become public at some point.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +184,30 @@ export function getInjectionTokens(element: Element): any[] {
-export function getDirectives(element: Element): {}[] {
-  const context = loadLContext(element)!;
+export function getDirectives(node: Node): {}[] {
+  if (node instanceof Text) {
```
```suggestion
  // Skip text node, since they can no host directives.
  if (node instanceof Text) {
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +184,30 @@ export function getInjectionTokens(element: Element): any[] {
+  if (node instanceof Text) {
+    return [];
+  }
+  const context = loadLContext(node, false);
+
+  if (context === null) {
+    return [];
+  }
```
May be add a quick note here?
```suggestion
  // Note: do not use `loadLContext` with `throwOnNotFound` argument defined as `true`,
  // as it might have performance implications in case throw happens. 
  const context = loadLContext(node, /* throwOnNotFound */ false);
  if (context === null) {
    return [];
  }
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +184,30 @@ export function getInjectionTokens(element: Element): any[] {
+  }
+
+  const tView = context.lView[TVIEW];
+  if (!tView || !tView.data[context.nodeIndex]) {
+    return [];
+  }
```
```suggestion
  if (!tView?.data[context.nodeIndex]) {
    return [];
  }
```
Though this should never happen and LView should always have associated TView (i.e. we can probably remove `?` here).

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -203,6 +218,61 @@ export function getDirectives(element: Element): {}[] {
+  outputs: Record<string, string>;
+}
+
+/**
+ * Metadata for a component developers can use for debugging purposes or tooling.
+ */
+export interface ComponentMetadata extends DirectiveMetadata {
+  encapsulation: ViewEncapsulation;
+  onPush: boolean;
+}
```
```suggestion
/**
 * Partial metadata for a given component instance.
 * This information might be useful for debugging purposes or tooling.
 * Currently the following fields are available:
 *  - inputs
 *  - outputs
 *  - encapsulation
 *  - changeDetection
 */
export interface ComponentDebugMetadata extends DirectiveDebugMetadata {
  encapsulation: ViewEncapsulation;
  changeDetection: ChangeDetectionStrategy;
}
```
Note: I'd propose align these fields closer to how we have them in the decorator, so I changed `onPush` to `changeDetection`

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -203,6 +218,61 @@ export function getDirectives(element: Element): {}[] {
   return context.directives === null ? [] : [...context.directives];
 }
 
+/**
+ * Metadata for a directive developers can use for debugging purposes or tooling.
+ */
+export interface DirectiveMetadata {
+  inputs: Record<string, string>;
+  outputs: Record<string, string>;
+}
```
I think we should add `Debug` to the interface names, so they are not confused with non-debug interfaces. Also it'd be great to mention that it's a partial metadata and only a certain subset of fields is avaialble.
```suggestion
/**
 * Partial metadata for a given directive instance.
 * This information might be useful for debugging purposes or tooling.
 * Currently only `inputs` and `outputs` metadata is available.
 */
export interface DirectiveDebugMetadata {
  inputs: Record<string, string>;
  outputs: Record<string, string>;
}
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -203,6 +218,61 @@ export function getDirectives(element: Element): {}[] {
+      inputs: componentDef.inputs,
+      outputs: componentDef.outputs,
+      encapsulation: componentDef.encapsulation,
+      onPush: componentDef.onPush
```
```suggestion
      // Align with the `ChangeDetectionStrategy` type
      onPush: componentDef.onPush ? 0 : 1
```

mgechev replied:
That's also the format we use in DevTools. Since we're not planning introducing other strategies, do you think it's ok keeping the way it is?

AndrewKushnir replied:
I think it'd still be great to align the value closer to the way it's present in decorator, since the `ng`-based functions can be used in other tools in the future as well.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -203,6 +218,61 @@ export function getDirectives(element: Element): {}[] {
+  onPush: boolean;
+}
+
+/**
+ * Returns the metadata for a particular directive or component. The function accepts an instance
+ * of a directive or component and returns the corresponding metadata.
+ *
+ * @param directiveOrComponentInstance Instance of a directive or component
+ * @returns metadata of the passed directive or component
+ */
+export function getDirectiveMetadata(directiveOrComponentInstance: any): ComponentMetadata|
+    DirectiveMetadata|null {
```
```suggestion
/**
 * Returns the debug (partial) metadata for a particular directive or component instance.
 * The function accepts an instance of a directive or component and returns the corresponding metadata.
 *
 * @param directiveOrComponentInstance Instance of a directive or component
 * @returns metadata of the passed directive or component
 */
export function getDirectiveDebugMetadata(directiveOrComponentInstance: {}): ComponentDebugMetadata|DirectiveDebugMetadata|null {
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -203,6 +218,61 @@ export function getDirectives(element: Element): {}[] {
+  return null;
+}
+
+/**
+ * Returns a `DebugNode` for a specified DOM node.
+ *
+ * @param node a DOM node the method will wrap in a DebugNode
+ * @returns a debug node
+ */
+export function asDebugNode(node: Node): PublicDebugNode {
+  return getDebugNode__POST_R3__(node);
+}
```
Should we create an instance of `DebugElement` (that has more methods) instead of `DebugNode`?
I'd also propose to rename the function from `asDebugNode` to `createDebugNode` - that might be a bit easier to read/understand.
```suggestion
/**
 * Creates an instance of `DebugNode` class for a specified DOM node.
 *
 * @see `DebugNode`
 *
 * @param node a DOM node that will be wrapped into a `DebugNode` instance
 * @returns an `DebugNode` instance
 */
export function createDebugNode(node: Node): DebugNode {
  return getDebugNode__POST_R3__(node);
}
```

mgechev commented on packages/core/src/core_render3_private_export.ts:
```diff
@@ -238,8 +233,17 @@ export {
-  ɵɵviewQuery,
-} from './render3/index';
+  ɵɵviewQuery} from './render3';
+export {
+  getLContext as ɵgetLContext
+} from './render3/context_discovery';
+export {
+  NG_COMP_DEF as ɵNG_COMP_DEF,
+  NG_DIR_DEF as ɵNG_DIR_DEF,
+  NG_ELEMENT_ID as ɵNG_ELEMENT_ID,
+  NG_MOD_DEF as ɵNG_MOD_DEF,
+  NG_PIPE_DEF as ɵNG_PIPE_DEF,
+} from './render3/fields';
```
That's where clang moves this export.

JoostK commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -203,11 +261,65 @@ export function getDirectives(element: Element): {}[] {
+  if (!constructor) {
+    throw new Error('Unable to find the instance constructor');
+  }
+  const directiveDef = getDirectiveDef(constructor);
+  if (directiveDef) {
+    return {inputs: directiveDef.inputs, outputs: directiveDef.outputs};
+  }
```
In case of a Component inheriting from a Directive I think this may result in the definition of the parent Directive, not the derived Component? (as `getDirectiveDef` does not use `hasOwnProperty`). It's typical that the component def is read first, e.g. in:

https://github.com/angular/angular/blob/bf158e7ff0715aabeae3c2c1ac923bf8cc7e4cfd/packages/core/src/render3/definition.ts#L367-L368

JoostK commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -7,21 +7,59 @@
+ * @returns If the element is a host element or not.
+ */
+function isHostElement(element: Element) {
+  const lViewOrLContext = readPatchedData(element);
+
+  // All host elements have LViews or LContexts associated with them.
+  // If we can't find patched data we can safely assume the current
+  // element is not a host.
+  if (lViewOrLContext === null) {
+    return false;
+  }
+  let lView: LView;
+  if (!Array.isArray(lViewOrLContext)) {
+    lView = lViewOrLContext.lView;
+  } else {
+    lView = lViewOrLContext;
+  }
+  if (!isLView(lView)) {
+    return false;
+  }
```
Could `readPatchedLView` be used here instead?

JoostK commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -264,7 +281,7 @@ export function getDirectiveMetadata(directiveOrComponentInstance: any): Compone
  * Throws if a given target doesn't have associated LContext.
  */
-export function loadLContext(target: {}): LContext;
+export function loadLContext(target: {}, throwOnNotFound?: boolean): LContext;
```
Oh and you might want to move this signature below the next one, as passing `false` would currently match this signature first and incorrectly have a non-nullable return type, I think?

mgechev replied:
The type checker correctly recognizes the type here, but it's fair to move it below.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +222,35 @@ export function getInjectionTokens(element: Element): any[] {
+    return [];
+  }
+
+  // Note: do not use `loadLContext` with `throwOnNotFound` argument defined as `true`,
+  // as it might have performance implications in case throw happens.
```
I think the comment should be moved to a place where the `getDirectives` is called, since it's now an argument value.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +222,35 @@ export function getInjectionTokens(element: Element): any[] {
  */
-export function getDirectives(element: Element): {}[] {
-  const context = loadLContext(element)!;
+export function getDirectives(node: Node, throwOnNotFound = true): {}[] {
```
```suggestion
export function getDirectives(node: Node, throwOnNotFound: boolean = true): {}[] {
```

AndrewKushnir replied:
Looking at other functions in `ng` namespace, they call `loadLContext` with `throwOnNotFound` as `false` and `getDirectives` does that differently. I believe that the flag was just missing here in the first place. So may be for v12 we should introduce a breaking change and always call `loadLContext` without triggering an error...

AndrewKushnir commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -47,6 +47,7 @@ export function publishDefaultGlobalUtils() {
      * The contract of the function might be changed in any release and/or the function can be
      * removed completely.
      */
+    publishGlobalUtil('getDirectiveMetadata', getDirectiveMetadata);
```
nit: may be move this 1 line below, so that the private method is the first one and all other getters are grouped.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +213,33 @@ export function getInjectionTokens(element: Element): any[] {
- * @param element DOM element for which to get the directives.
- * @returns Array of directives associated with the element.
+ * @param node DOM node for which to get the directives.
+ * @param throwOnNotFound boolean flag indicating if we should throw an error in case we can't
```
This param was removed, let's remove it form the docs too.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -6,22 +6,51 @@
+  if (!isLView(lView)) {
+    return false;
+  }
+  const tView = lView[TVIEW];
+  const components = tView.components;
+  return components?.some(component => {
+    const tNode = tView.data[component];
+    if (!tNode || !isTNode(tNode) || lView[component][HOST] !== element) {
+      return false;
```
I'd propose moving `tView.data` outside of the loop to avoid extra property reads.
```suggestion
  const tView = lView[TVIEW];
  const tData = tView.data;
  const components = tView.components;
  return components?.some(component => {
    const tNode = tData[component];
    if (!tNode || !isTNode(tNode) || lView[component][HOST] !== element) {
      return false;
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,18 +213,33 @@ export function getInjectionTokens(element: Element): any[] {
+    return [];
+  }
+
+  const tView = context.lView[TVIEW];
+  if (!tView?.data[context.nodeIndex]) {
+    return [];
+  }
 
   if (context.directives === undefined) {
     context.directives = getDirectivesAtNodeIndex(context.nodeIndex, context.lView, false);
```
We can also extract `lView` and `nodeIndex` reads here:
```suggestion
  const lView = context.lView;
  const tView = lView[TVIEW];
  const nodeIndex = context.nodeIndex;
  if (!tView?.data[nodeIndex]) {
    return [];
  }

  if (context.directives === undefined) {
    context.directives = getDirectivesAtNodeIndex(nodeIndex, lView, false);
```

AndrewKushnir reviewed (approved):
Looks great, thanks for addressing the feedback @mgechev 👍 

FYI, I've added the "breaking changes" label to indicate that the `ng.getDirectives` function never throws now. Could you please add the "BREAKING CHANGE" note to a commit that adds `false` to the `loadLContext` call?

AndrewKushnir commented:
[Presubmit](http://test/OCL:367752282:BASE:367752289:1618029551253:19af695b).

crisbeto commented on packages/core/src/render3/interfaces/type_checks.ts:
```diff
@@ -48,3 +48,7 @@ export function isComponentDef<T>(def: DirectiveDef<T>): def is ComponentDef<T>
 }
+
+export function isTNode(obj: unknown): obj is TNode {
+  return !!obj && typeof obj === 'object' && (obj as {}).hasOwnProperty('directiveStylingLast');
```
Isn't this going to break for apps using property minification (e.g. Closure)?

petebacondarwin reviewed (commented):
The first commit message doesn't seem right:

> This commit introduces two global debugging method `ng.getDirectiveMetadata` which returns the metadata for a directive or component instance.

I think there is only one new method right?

petebacondarwin reviewed (commented):
The second commit doesn't actually appear to change any behaviour?

petebacondarwin commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -11,20 +11,47 @@ import {Injector} from '../../di/injector';
+  const tView = lView[TVIEW];
+  const tData = tView.data;
+  const components = tView.components;
+  return components?.some(component => {
+    const tNode = tData[component];
+    if (!tNode || !isTNode(tNode) || lView[component][HOST] !== element) {
+      return false;
+    }
+    return isComponentHost(tNode);
+  }) ??
+      false;
```
It may be the fault of the formatter but I found this statement hard to grok. How about:

```ts
if (components === undefined) {
  return false;
}
return components.some(component => {
  const tNode = tData[component];
  return (!tNode || !isTNode(tNode) || lView[component][HOST] !== element) ? false : isComponentHost(tNode);
});
```

petebacondarwin commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -6,22 +6,52 @@
+  const components = tView.components;
+  return components?.some(component => {
+    const tNode = tData[component];
+    if (!tNode || !isTNode(tNode) || lView[component][HOST] !== element) {
```
`isTNode()` already checks whether `tNode` is truthy.

```suggestion
    if (!isTNode(tNode) || lView[component][HOST] !== element) {
```

AndrewKushnir commented:
@mgechev FYI, the presubmit indicated a few targets broken by this change (most likely with the same root cause), see presubmit results [here](http://test/OCL:367752282:BASE:367752289:1618029551253:19af695b).

jessicajaniuk reviewed (approved):
LGTM aside from the various feedbacks :cookie:

reviewed-for: public-api

AndrewKushnir commented:
Just a quick update after failing targets investigation. The problem was caused by the missing `unwrapRNode` call here (the `lView[component][HOST]` may also point to an LView, not always an RNode):

```
    if (!tNode || !isTNode(tNode) ||
        unwrapRNode(lView[component]) !== element) { // <-- `unwrapRNode` was missing here
      return false;
    }
```

However we'd need to add more tests to make sure this change is completely safe.

The proposal is to reduce the scope of this PR to just extra API calls added to the `ng` namespace (and adding `false` to the `loadLContext` call), so we can land it sooner and followup with the `isComponentHostElement` perf optimization in a separate PR. 

Thank you.

mgechev committed:
perf(core): optimize getDirectives

This commit introduces the following optimizations:

1. We return an empty array for text nodes in `getDirectives` because
Angular does not support attaching logic to them. This optimization
improves performance of `getDirectives` significantly because text nodes
often result in expensive calls to `loadLContext` since we can't resolve
it from a parent node.
1. `getDirectives` now calls `loadLContext` with second argument `false`
so it doesn't throw an error. This brings another significant
improvement because prevents the VM from deoptimizing calls.

BREAKING CHANGE:

Previously the `ng.getDirectives` function threw an error in case a
given DOM node had no Angular context associated with it (for example
if a function was called for a DOM element outside of an Angular app).
This behavior was inconsistent with other debugging utilities under `ng`
namespace, which handled this situation without raising an exception.
Now calling the `ng.getDirectives` function for such DOM nodes would
result in an empty array returned from that function.

AndrewKushnir reviewed (approved):
LGTM, thanks @mgechev 👍

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:367752282:BASE:368114855:1618272446261:b448637a).

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
@mgechev FYI presubmits are successful for the changes in this PR  🎉

petebacondarwin commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,28 +184,99 @@ export function getInjectionTokens(element: Element): any[] {
+ *
+ * @param directiveOrComponentInstance Instance of a directive or component
+ * @returns metadata of the passed directive or component
+ */
```
Given that this is being published via `publishGlobalUtils()` then I would expect this to appear in our docs?

```suggestion
 * @publicApi
 * @globalApi ng
 */
```

petebacondarwin commented on packages/core/test/acceptance/discover_utils_spec.ts:
```diff
@@ -359,6 +382,12 @@ onlyInIvy('Ivy-specific utilities').describe('discovery utils deprecated', () =>
+    it('should not throw if it cannot find LContext', () => {
+      expect(() => {
+        getDirectives(document.createElement('div'));
+      }).not.toThrow();
```
Would it actually be worth asserting what the return value of such a call is? Is it `[]`?

petebacondarwin commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -181,28 +184,99 @@ export function getInjectionTokens(element: Element): any[] {
+ * @param directiveOrComponentInstance Instance of a directive or component
+ * @returns metadata of the passed directive or component
+ */
+export function getDirectiveMetadata(directiveOrComponentInstance: any): ComponentMetadata|
+    DirectiveMetadata|null {
```
And if this does appear in the public docs, should the two interfaces `ComponentMetadata` and `DirectiveMetadata` also be published to the docs?

petebacondarwin reviewed (changes requested):
The first commit is not accurate:

```
This commit introduces two global debugging method
`ng.getDirectiveMetadata` which returns the metadata for a directive or
component instance.
```

Only one new utility is added.

We need to add jsdoc tags to this new utility and the types it returns for them to appear in the docs.

mary-poppins commented:
You can preview 6505ad1 at https://pr41525-6505ad1.ngbuilds.io/.
You can preview aeb978c at https://pr41525-aeb978c.ngbuilds.io/.
You can preview 1c197ff at https://pr41525-1c197ff.ngbuilds.io/.
You can preview fb8943e at https://pr41525-fb8943e.ngbuilds.io/.
You can preview ab83db9 at https://pr41525-ab83db9.ngbuilds.io/.
You can preview 145bf47 at https://pr41525-145bf47.ngbuilds.io/.
You can preview f5c7d9c at https://pr41525-f5c7d9c.ngbuilds.io/.
You can preview 0547c28 at https://pr41525-0547c28.ngbuilds.io/.

mary-poppins commented:
You can preview a435b20 at https://pr41525-a435b20.ngbuilds.io/.

mary-poppins commented:
You can preview f441542 at https://pr41525-f441542.ngbuilds.io/.

petebacondarwin reviewed (approved):
Reviewed-for: public-api

mary-poppins commented:
You can preview 90b4bf5 at https://pr41525-90b4bf5.ngbuilds.io/.

petebacondarwin commented:
The aio-test failure is due to this line: https://github.com/angular/angular/blob/8c939dcb40ea12dfa8c3e5de0cb04d33f0a1e6f4/aio/ngsw-config.json#L82

mary-poppins commented:
You can preview a297877 at https://pr41525-a297877.ngbuilds.io/.

petebacondarwin commented on packages/core/src/render3/index.ts:
```diff
@@ -176,12 +176,15 @@ export { ɵɵtemplateRefExtractor} from './view_engine_compatibility_prebound';
 // clang-format on
 
 export {
+  ComponentDebugMetadata as ComponentMetadata,
```
I don't think this (and the directive on below) is correct. We should not be aliasing these anymore.

petebacondarwin replied:
Which is why they are appearing incorrectly in the API reference documentation: https://pr41525-a297877.ngbuilds.io/api/core/global#structures and are not clickable in the https://pr41525-a297877.ngbuilds.io/api/core/global/ngGetDirectiveMetadata doc

mgechev replied:
Yes, you're right. Code's refactoring introduced it. Removing them now.

mary-poppins commented:
You can preview 9b60ae4 at https://pr41525-9b60ae4.ngbuilds.io/.

petebacondarwin reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:367752282:BASE:368303843:1618351672380:29b1e394).
